src/main/java/duke/Parser.java

```diff
@@ -0,0 +1,187 @@
+package duke;
+
+import java.time.LocalDateTime;
+import java.time.format.DateTimeFormatter;
+import java.time.format.DateTimeParseException;
+
+/**
+ * This class encapsulates the mechanism to parse user commands.
+ *
+ * @author Kleon Ang
+ */
+public class Parser {
+    private static final DateTimeFormatter FORMATTER = DateTimeFormatter.ofPattern("yyyy-MM-d H:mm");
+    private final TaskList tasks;
+    private boolean toRewriteData;
+    private boolean isExit;
+    private boolean isFind;
+
+    /**
+     * Constructor for a Parser class.
+     *
+     * @param tasks A TaskList for the Parser to refer to.
+     */
+    public Parser(TaskList tasks) {
+        this.tasks = tasks;
+        this.toRewriteData = false;
+        this.isExit = false;
+        this.isFind = false;
+    }
+
+    private void list(TaskList tasks) throws DukeException {
+        if (tasks.size() == 0) {
+            throw new DukeException("There are currently no tasks in your list.");
+        }
+        String matching = this.isFind ? "matching " : "";
+        StringBuilder tasksBuilder = new StringBuilder();
+        tasksBuilder.append("Here are the ").append(matching).append("tasks in your list:\n");
+        for (int i = 0; i < tasks.size(); ++i) {
+            String counter = String.valueOf(i + 1);
+            Task currentTask = tasks.get(i);
+            tasksBuilder.append(counter).append(".").append(currentTask);
+            if (i < tasks.size() - 1) {
+                // Append newline for all tasks before last task
+                tasksBuilder.append("\n");
+            }
+        }
+        Ui.printReply(tasksBuilder.toString());
+    }
+
+    private void done(int counter) throws DukeException {
```

Comment:
"counter" varName might be a bit ambiguous, perhaps "TaskIndex" would be more intuitive